lib/kcl.js

```diff
@@ -215,19 +218,23 @@ module.exports = function(config, kinesis) {
         throttledPutToCloudwatch('ShardIteratorAgeInMs', (+new Date()) - resp.Records[0].ApproximateArrivalTimestamp, 'Milliseconds', shard.id);
       }
 
-      // malformed responses will have an undefined NextShardIterator. In that case, we keep the existing one.
-      shard.iterator = resp.data.NextShardIterator === undefined ? shard.iterator : resp.data.NextShardIterator;
+      // malformed responses will have an undefined NextShardIterator and undefined Records key
+      // only replace the iterator if we have a nextShardIterator. otherwise we'll keep using it
+      if (resp.data.NextShardIterator) shard.iterator = resp.data.NextShardIterator;
       shard.lastGetRecords = +new Date();
 
-      if (resp.data.Records && resp.data.Records.length > 0) {
+      if (resp.data.Records && resp.data.Records.length > 0 && shard.iterator) {
```

Comment:
What are you checking for by adding `&& shard.iterator`

Even if the request had undefined or null iterator, `shard.iterator` will still have the previous value.